### PR TITLE
WebGLRenderer: Add mipmap & compressed texture handling to .copyTextureToTexture().

### LINKED ISSUE #14423: copyTextureToTexture(): Wrong processing of mipmaps
##### Description of the problem

There is something seriously wrong with the dataTexture and copyTextureToTexture functions. 
I was struggling for many days with this issue until I was convinced that this is a three.js bug.

When I upload an image to texture1 and then copy that texture to a datatexture texture2, I cannot use the minFilter and magFilters for mipmaps at all (LinearMipmapLinearFilter and LinearFilter correspondingly) for the destination datatexture, as the texture will not be rendered, so the only option is to enjoy wild aliasing artifacts when applied to a moving plane or other object.

In order to make sure that this is a three.js bug, I added the WEBGL equivalent functions on an example WEBGL code and I was able to use mipmaps and filtering on the destination data texture with no problems at all.

>You can see the unfiltered datatexture artifacts on stackoverflow:
https://stackoverflow.com/questions/51071390/copytexturetotexture-results-in-horrible-aliasing-artifacts-in-three-js

>Or better try to mod this jsfiddle. You could try to uncomment and include the lines 73-76 that enable mipmaps and filtering.
https://jsfiddle.net/Debanker/4gL0eosh/27/

>You can also see the properly working WEBGL version that I mentioned, where the same image is applied on a cube (instead of a plane), but the two required helper scripts (on the top of HTML) are blocked (otherwise it works offline -unless you can host those files elsewhere and/or you can make it work -I spent half a day to find a proper free host to no avail). You can change line 326 from rendering texture2(copied) to texture1(original).
https://jsfiddle.net/Debanker/sybtL7ed/24/

##### Three.js version

- [ ] Dev
- [x] r93
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [ ] All of them
- [x] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

### LINKED ISSUE #18736: WebGLRenderer.copyTextureToTexture() does not automatically generate mipmaps
##### Description of the problem

WebGLRenderer.copyTextureToTexture() does not automatically generate new mipmaps (from the full size texture) when Texture.generateMipmaps == true) 

Repro steps:
Create a texture
Set .generateMipmaps = true
Copy a smaller texture to it with WebGLRenderer.copyTextureToTexture()
When the camera gets further away, the region of the texture you copied 'fades out' as the smaller (unchanged...) mipmap gets sampled.

See:
https://frankolivier.github.io/bugs/copyTextureToTexture-and-mipmaps/before/
(Notice how the green tile fades out)

Correct behavior:
https://frankolivier.github.io/bugs/copyTextureToTexture-and-mipmaps/after/

##### Three.js version

- [X] Dev
- [X] r113

##### Browser

- [x] All of them

##### OS

- [x] All of them

### LINKED ISSUE #18821: WebGLRenderer.copyTextureToTexture() does not support compressed textures
##### Description of the problem

`WebGLRenderer.copyTextureToTexture()` does not support compressed textures

When using regular PNGs, I can use `WebGLRenderer.copyTextureToTexture()` to add a green tile to the face of a rotating cube:
https://frankolivier.github.io/bugs/copyTextureToTexture-and-compressed-textures/png/

When using (Basis) compressed textures, it does not work:
https://frankolivier.github.io/bugs/copyTextureToTexture-and-compressed-textures/basis-broken
(Observe: no green tile)

The fix is to add 

`if ( srcTexture.isCompressedTexture ) {`
`_gl.compressedTexSubImage2D( 3553, level, position.x, position.y, srcTexture.mipmaps[0].width, srcTexture.mipmaps[0].height, glFormat, srcTexture.mipmaps[0].data );`
`}`

...to `.copyTextureToTexture()`

See a fixed example in https://frankolivier.github.io/bugs/copyTextureToTexture-and-compressed-textures/basis-fixed/

##### Three.js version

- [x] Dev
- [x] r114
- [ ] ...

##### Browser
- [x] All of them

##### OS
- [x] All of them

##### Hardware Requirements
Any

### PR DESCRIPTION
**Pull request by frankolivier**
Fixes #18736 
Fixes #18821

### CODE REVIEW

File: src/renderers/WebGLRenderer.js [hunk 1]
```diff
@@ -2820,6 +2820,8 @@ function WebGLRenderer( parameters ) {
 
 		}
 
+		if (dstTexture.generateMipmaps) _gl.generateMipmap( _gl.TEXTURE_2D );
```
Comment (Mugen87) [inline src/renderers/WebGLRenderer.js]: Instead of generating them, is it possible to copy the existing ones?

Comment (Mugen87) [inline src/renderers/WebGLRenderer.js]: I'm not sure this check is safe. What if the flag is set to `false` since mipmaps were defined manually?

**Comment by frankolivier** [comment]
Added a check for mipmap level 0; added 0 as the default level for this function

IMO the check is safe - if the developer wanted to manually handle mipmaps ( by setting dstTexture.generateMipmaps = `false` ) _gl.generateMipmap() should not be called in the added code.

File: src/renderers/WebGLRenderer.js [hunk 2]
```diff
@@ -2801,7 +2801,7 @@ function WebGLRenderer( parameters ) {
 
 	};
 
-	this.copyTextureToTexture = function ( position, srcTexture, dstTexture, level ) {
+	this.copyTextureToTexture = function ( position, srcTexture, dstTexture, level = 0 ) {
```
Comment (donmccurdy) [inline src/renderers/WebGLRenderer.js]: Perhaps we're inevitably headed toward using ES6, but for now there are no occurrences of default parameters in the library (that I know of?) and our Rollup build does not do a general-purpose ES6->ES5 transpile. Until an explicit decision is made to start using ES6, I think it'd be best to do an old-fashioned `level = level || 0` instead here.

File: src/renderers/WebGLRenderer.js [hunk 3]
```diff
@@ -2803,6 +2803,8 @@ function WebGLRenderer( parameters ) {
 
 	this.copyTextureToTexture = function ( position, srcTexture, dstTexture, level ) {
 
+		level = level || 0;
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: I guess this makes it more clear?
```js
if ( level === undefined ) level = 0;
```

File: src/renderers/WebGLRenderer.js [hunk 4]
```diff
@@ -2812,14 +2814,16 @@ function WebGLRenderer( parameters ) {
...
+			_gl.texSubImage2D( _gl.TEXTURE_2D, level, position.x, position.y, width, height, glFormat, glType, srcTexture.image.data );
 
 		} else {
 
-			_gl.texSubImage2D( _gl.TEXTURE_2D, level || 0, position.x, position.y, glFormat, glType, srcTexture.image );
+			_gl.texSubImage2D( _gl.TEXTURE_2D, level, position.x, position.y, glFormat, glType, srcTexture.image );
 
 		}
 
+		if ( level == 0 && dstTexture.generateMipmaps ) _gl.generateMipmap( _gl.TEXTURE_2D );
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: I am confused now...

Is this the intention?

```js
if ( level === undefined && dstTexture.generateMipmaps ) _gl.generateMipmap( _gl.TEXTURE_2D );
```

File: src/renderers/WebGLRenderer.js (hunk 3 above)
Comment (frankolivier) [reply src/renderers/WebGLRenderer.js]: Will change to `if ( level === undefined ) level = 0;`

File: src/renderers/WebGLRenderer.js (hunk 4 above)
Comment (frankolivier) [reply src/renderers/WebGLRenderer.js]: No... the intention is only to autogenerate mipmaps if the developer is updating the level 0 texture, and when the generateMipmaps flag is set.

Comment (mrdoob) [reply src/renderers/WebGLRenderer.js]: Ah! Okay.

Do you mind adding a comment?

```js
// Generate mipmaps only when copying level 0
```

Comment (frankolivier) [reply src/renderers/WebGLRenderer.js]: Added

**Comment by Mugen87** [comment]
Can you please remove `build/three.module.js ` from your PR?

**Comment by Mugen87** [comment]
Um, now you have completely removed `build/three.module.js`. The idea was to just undo your changes to `build/three.module.js`.

**Review by Mugen87 — APPROVED** [review]

**Comment by mrdoob** [comment]
> Fixes #14423

Hmm, do you mind explaining how this fixes #14423?

**Comment by frankolivier** [comment]
Ah, my bad - it would not actually fix #14423. 

I've filed a new issue #18736 for the actual issue I was trying to fix for my project.

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**